prow/config/config.go

```diff
@@ -693,6 +694,121 @@ type Deck struct {
 	RerunAuthConfigs RerunAuthConfigs `json:"rerun_auth_configs,omitempty"`
+	// RestrictStoragePaths restricts artifact requests to specific buckets.
+	// By default, buckets listed in the GCSConfiguration are automatically allowed.
+	// Additional locations can be allowed via `AdditionalAllowedBuckets` fields.
+	// When unspecified (nil), it defaults to false (until ~Jan 2021).
+	RestrictStoragePaths *bool `json:"restrict_storage_paths,omitempty"`
+	// AdditionalAllowedBuckets is a list of storage buckets to allow in artifact requests
+	// (in addition to those listed in the GCSConfiguration).
+	// Setting this field requires "RestrictStoragePaths" also be set to `true`.
+	AdditionalAllowedBuckets []string `json:"additional_allowed_buckets,omitempty"`
+	// AllKnownStorageBuckets contains all storage buckets configured in all of the
+	// job configs.
+	AllKnownStorageBuckets sets.String `json:"-"`
+}
+
+// Validate performs validation and sanitization on the Deck object.
+func (d *Deck) Validate() error {
+	if len(d.AdditionalAllowedBuckets) > 0 && !d.ShouldValidateStorageBuckets() {
+		return fmt.Errorf("deck.restrict_storage_paths is not enabled despite deck.additional_allowed_buckets being configured: %v", d.AdditionalAllowedBuckets)
+	}
+
+	// TODO(@clarketm): Remove "rerun_auth_config" validation in July 2020
+	if d.RerunAuthConfig != nil {
+		logrus.Warning("rerun_auth_config will be deprecated in July 2020, and it will be replaced with rerun_auth_configs['*'].")
+
+		if d.RerunAuthConfigs != nil {
+			return errors.New("rerun_auth_config and rerun_auth_configs['*'] are mutually exclusive")
+		}
+
+		d.RerunAuthConfigs = RerunAuthConfigs{"*": *d.RerunAuthConfig}
+	}
+
+	// Note: The RerunAuthConfigs logic isn't deprecated, only the above RerunAuthConfig stuff is
+	if d.RerunAuthConfigs != nil {
+		for k, config := range d.RerunAuthConfigs {
+			if err := config.Validate(); err != nil {
+				return fmt.Errorf("rerun_auth_configs[%s]: %v", k, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+var warnInRepoStorageBucketValidation time.Time
+
+// ValidateStorageBucket validates a storage bucket, *if* the `Deck.RestrictStoragePaths` config field is true.
+// The given storage path should match "<key-type>/<bucket>/<folders...>".
+// The bucket name should match either the built-in and default storage locations _or_
+// should be contained in the `Deck.AdditionalAllowedBuckets` config field.
+// If the `Deck.RestrictStoragePaths` field is false, no validation errors are returned.
+func (c *Config) ValidateStorageBucket(path string) error {
+	parts := strings.Split(path, "/")
+	if len(parts) < 3 {
+		return fmt.Errorf("invalid path: %s expected <key-type>/<bucket>/<folders...>", path)
+	}
+
+	if len(c.InRepoConfig.Enabled) > 0 && len(c.Deck.AdditionalAllowedBuckets) == 0 {
+		logrusutil.ThrottledWarnf(&warnInRepoStorageBucketValidation, 1*time.Hour, strings.Join([]string{
```

Comment:
nit: You don't need to use `strings.Join` here. You can
1. use ``` ` ``` for multiline raw strings
1. concatenate lines with `+` (the compiler should be smart enough to do this for literals at compile time rather than runtime)
1. use a long single line literal since this is for an error message.

I think those options are more efficient and more readable.

Reply:
Done.